crates/egui/src/atomic_layout.rs

```diff
@@ -0,0 +1,740 @@
+            gap,
+        }
+    }
+}
+
+/// Instructions for painting an [`AtomicLayout`].
+#[derive(Clone, Debug)]
+pub struct AllocatedAtomicLayout<'a> {
+    pub sized_atomics: Vec<SizedAtomic<'a>>,
+    pub frame: Frame,
+    pub fallback_text_color: Color32,
+    pub response: Response,
+    grow_count: usize,
+    // The size of the inner content, before any growing.
+    desired_size: Vec2,
+    align2: Align2,
+    gap: f32,
+}
+
+impl<'a> AllocatedAtomicLayout<'a> {
+    /// Paint the [`Frame`] and individual [`Atomic`]s.
+    pub fn paint(self, ui: &Ui) -> AtomicLayoutResponse {
+        let Self {
+            sized_atomics: sized_items,
+            frame,
+            fallback_text_color,
+            response,
+            grow_count,
+            desired_size,
+            align2,
+            gap,
+        } = self;
+
+        let inner_rect = response.rect - self.frame.total_margin();
+
+        ui.painter().add(frame.paint(inner_rect));
+
+        let width_to_fill = inner_rect.width();
+        let extra_space = f32::max(width_to_fill - desired_size.x, 0.0);
+        let grow_width = f32::max(extra_space / grow_count as f32, 0.0);
+
+        let aligned_rect = if grow_count > 0 {
+            align2.align_size_within_rect(Vec2::new(width_to_fill, desired_size.y), inner_rect)
+        } else {
+            align2.align_size_within_rect(desired_size, inner_rect)
+        };
+
+        let mut cursor = aligned_rect.left();
+
+        let mut response = AtomicLayoutResponse {
+            response,
+            custom_rects: HashMap::new(),
+        };
+
+        for sized in sized_items {
+            let size = sized.size;
+            let growth = if sized.grow { grow_width } else { 0.0 };
+
+            let frame = aligned_rect
+                .with_min_x(cursor)
+                .with_max_x(cursor + size.x + growth);
+            cursor = frame.right() + gap;
+
+            let align = Align2::CENTER_CENTER;
+            let rect = align.align_size_within_rect(size, frame);
+
+            match sized.kind {
+                SizedAtomicKind::Text(galley) => {
+                    ui.painter().galley(rect.min, galley, fallback_text_color);
+                }
+                SizedAtomicKind::Image(image, _) => {
+                    image.paint_at(ui, rect);
+                }
+                SizedAtomicKind::Custom(id, _) => {
+                    response.custom_rects.insert(id, rect);
+                }
+                SizedAtomicKind::Empty => {}
+            }
+        }
+
+        response
+    }
+}
+
+/// Response from a [`AtomicLayout::show`] or [`AllocatedAtomicLayout::paint`].
+///
+/// Use the `custom_rects` together with [`AtomicKind::Custom`] to add child widgets to a widget.
+///
+/// NOTE: Don't `unwrap` rects, they might be empty when the widget is not visible.
+#[derive(Clone, Debug)]
+pub struct AtomicLayoutResponse {
+    pub response: Response,
+    pub custom_rects: HashMap<Id, Rect>,
```

Comment:
This will cause additional allocation for each atomic. We could avoid this by instead using `SmallVec<[(Id, Rect); 2]>`

Reply:
…with a helper for doing the lookups